GoogleForCreators/web-stories-wp
# Tenor: Render as video rather than gif

## Summary

- Provide first-class output component for `gif` media from Tenor so it renders as a video

## Relevant Technical Choices

Quotes are details from the story, which is linked at the bottom of this PR. 

> The Media3P API Returns both mp4 and webm formats, you probably want to just use the mp4s for now.
> The easiest approach is possibly to have the canvas (and design panel) treat the GIF as an image, and then at the time of rendering the element, render it as a video with autoplay, mute and no controls.

- The above is the approach I took. For the element's `resource` I included a new property called `output` with the video mimeType (`video/mp4`) and a `src` (a link to the full sized mp4), as well as `sizes.mp4` and `sizes.webm` in case we would like to support both sources in the future. Also open to changing `elements/gif/output` to support both by default.

> As I mentioned previously, I think the cleanest way to achieve this would be by making gif a first-class citizen here.
> To give an example, we currently have a couple of instances like this in the code base:
```
if (resource.type == 'gif') {
  resource.type = 'image';
}
```
> Adding more checks like this is going to make things overly complex and prone to errors (should we ever forget to add such a check anywhere).

- With the above in consideration, I created a new folder `gif` which leverages the `image` element type in its `index.js` but then renders a new component for output: `gif/output`. I removed modifications to `resource.type` in `useInsertElement` and `mediaElement` to treat gifs as first-class citizens, and changed a check in `media/edit` to treat the `gif` type as an image.

- gifs uploaded from local have `resource.type = 'image'` so no modifications were required to render those properly

- I also added 'gif' as a drop target in the `dropTargets/provider` which allows it to be masked by shapes like videos and images!
<img width="437" alt="Screen Shot 2020-10-15 at 11 37 13 AM" src="https://user-images.githubusercontent.com/41136059/96172152-d38fad00-0eda-11eb-9dc8-60b0a83a2534.png">

<img width="492" alt="Screen Shot 2020-10-15 at 11 38 07 AM" src="https://user-images.githubusercontent.com/41136059/96172213-edc98b00-0eda-11eb-98c4-2cfd75739e0c.png">


## To-do

- [x] `elements/gif/test` it should produce valid AMP output
- [x] `elements/gif/test` it should produce an AMP video with autoplay, no controls, no audio, and loop
- [x] `app/media/utils/test` `getResourceFromMedia3P` should return a valid shape for `gif` type resources
- [x] ~~get a rundown on Karma so I can better determine if an e2e test would be appropriate here. I don't think so, but I'd love to learn something new. 🧠~~
- [x] move modules shared by image and gif to media directory
- [x] replace string matchers in tests with snapshot

__finishing touches 💅__
- [x] add `sizes` to gif prop-types
- [x] use the abstract `media/*` elements for elements/gif the same way elements/image is using them rather than straight up re-using the image/element

## User-facing changes

There should be no perceptible difference to the user from these changes.

## Testing Instructions

1. Open a new story in the editor
2. Go to the images tab and click on "GIFs"
3. Double click a gif to add it to the editor
4. Click Preview to preview the story
5. Right-click on the gif in the preview pane and select "Inspect Element in your browser's context menu
6. The gif should be rendered as an `<amp-video />` element e.g.
<img width="1023" alt="Screen Shot 2020-10-12 at 3 20 05 PM" src="https://user-images.githubusercontent.com/41136059/95795130-78717680-0c9e-11eb-8db3-9e4acee3919b.png">

---

Fixes #4782
